### ISSUE TITLE
How to fine tune yolov8 model again with Focal Loss ?

### ISSUE THREAD

**Issue by Abhijeet241093** [ultralytics/ultralytics#19979]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

How to fine tune yolov8 model again with Focal Loss ? for better performance

### Additional

Thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Abhijeet241093, thank you for your question and for being part of the Ultralytics community 🚀!

To help you with fine-tuning your YOLO model using Focal Loss, we recommend exploring the [Ultralytics Docs](https://docs.ultralytics.com/). The [Python usage guide](https://docs.ultralytics.com/usage/python/) and [custom training tips](https://docs.ultralytics.com/guides/model-training-tips/) are excellent starting points for customizing your training processes.

If this is a ❓ Question about modifying the loss function, please share more details about your implementation, such as the changes you've made, your dataset, and any logs or errors encountered. This will help us provide more targeted guidance.

If this is a 🐛 Bug Report related to your modifications, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can debug effectively.

## Upgrade

Ensure that you're using the latest version of the `ultralytics` package to access the newest features and fixes. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment with:

```bash
pip install -U ultralytics
```

## Community and Support

For further discussions or help, join the Ultralytics community:
- Real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions on [Discourse](https://community.ultralytics.com/)
- Share ideas and knowledge on our [Subreddit](https://reddit.com/r/Ultralytics)

## Verified Environments

YOLO models, including YOLO8, can be run in any of these environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests ensure correct performance across YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu.

This is an automated response, but no worries—an Ultralytics engineer will review your issue and assist you soon 😊!

**Comment by glenn-jocher** [comment]
@Abhijeet241093 to implement Focal Loss with YOLOv8 during fine-tuning, you can modify your training configuration to use the built-in `FocalLoss` class from the Ultralytics utils. Here's a concise implementation approach:

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')  # Load pretrained
model.train(data='your_data.yaml', loss='focal', ...)  # Add other params
```

The Focal Loss implementation is available in our [ultralytics/utils/loss.py](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.FocalLoss) reference. For optimal results, consider adjusting the `alpha` and `gamma` parameters in your hyperparameter file (hyp.yaml) while monitoring validation metrics.

**Comment by gongjizhang** [comment]
![Image](https://github.com/user-attachments/assets/bb82efd6-a737-4141-b956-b1ef1c534c24)

**Comment by glenn-jocher** [comment]
To implement Focal Loss with YOLOv8, you can directly specify `loss='focal'` in your training configuration as shown in our [loss.py reference](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.FocalLoss). For optimal results, we recommend experimenting with the `gamma` and `alpha` parameters in your hyperparameters file (hyp.yaml) while monitoring validation metrics. The Ultralytics team appreciates your engagement with YOLO! 🚀

**Comment by CyrusAddy** [comment]
When I try to ran the loss='focal' I get same error as above.

**Comment by glenn-jocher** [comment]
I see the issue now. The direct `loss='focal'` parameter isn't currently supported in the public API. Instead, you'll need to create a custom trainer that implements Focal Loss. Here's a more accurate approach:

```python
from ultralytics import YOLO
from ultralytics.utils.loss import FocalLoss
from ultralytics.engine.trainer import BaseTrainer
from torch import nn

# Custom trainer that uses FocalLoss
class FocalLossTrainer(BaseTrainer):
    def get_criterion(self):
        criterion = super().get_criterion()
        # Replace the classification loss component with FocalLoss
        # Keep gamma and alpha as parameters you can tune
        criterion.bce = FocalLoss()
        return criterion

# Load model and train with custom trainer
model = YOLO('yolov8n.pt')
results = model.train(data='your_data.yaml', trainer=FocalLossTrainer)
```

This approach directly modifies the loss function by extending the trainer class. You can adjust the gamma and alpha parameters within the FocalLoss initialization if needed.
